src/VisualStudio/Core/Def/Implementation/ColorSchemes/ColorSchemeApplier.cs

```diff
@@ -0,0 +1,280 @@
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Utilities;
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.VisualStudio.LanguageServices.ColorSchemes
+{
+    [Export(typeof(IWpfTextViewConnectionListener))]
+    [Export(typeof(ColorSchemeApplier))]
+    [ContentType(ContentTypeNames.RoslynContentType)]
+    [TextViewRole(PredefinedTextViewRoles.Analyzable)]
+    internal sealed partial class ColorSchemeApplier : ForegroundThreadAffinitizedObject, IWpfTextViewConnectionListener, IDisposable
+    {
+        private const string RoslynTextEditorRegistryKey = "Themes\\{de3dbbcd-f642-433c-8353-8f1df4370aba}\\Roslyn Text Editor MEF Items";
+        private const string UseEnhancedColorsSetting = "WindowManagement.Options.UseEnhancedColorsForManagedLanguages";
+        private const string AppliedColorSchemeKeyName = "ColorScheme";
+
+        private readonly IServiceProvider _serviceProvider;
+        private readonly ForegroundColorDefaulter _colorApplier;
+        private readonly ISettingsManager _settingsManager;
+
+        private bool _isDisposed = false;
+        private bool isInitialized = false;
+
+        [ImportingConstructor]
+        [Obsolete]
+        public ColorSchemeApplier(IThreadingContext threadingContext, [Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider)
+            : base(threadingContext)
+        {
+            _serviceProvider = serviceProvider;
+            _settingsManager = (ISettingsManager)_serviceProvider.GetService(typeof(SVsSettingsPersistenceManager));
+            _colorApplier = new ForegroundColorDefaulter(threadingContext, _serviceProvider);
+        }
+
+        public void Dispose()
+        {
+            // Dispose is invoked when the MEF container is disposed. This will be our
+            // signal that VS is shutting down and we shouldn't try and perform any work.
+            _isDisposed = true;
+        }
+
+        public void SubjectBuffersConnected(IWpfTextView textView, ConnectionReason reason, Collection<ITextBuffer> subjectBuffers)
+        {
+            AssertIsForeground();
+
+            if (!isInitialized)
+            {
+                isInitialized = true;
+
+                // We need to update the theme whenever the Editor Color Scheme setting changes or the VS Theme changes.
+                _settingsManager.GetSubset(ColorSchemeOptions.SettingKey).SettingChangedAsync += ColorSchemeChanged;
+                VSColorTheme.ThemeChanged += VSColorTheme_ThemeChanged;
+
+                // Check on first connect whether we need to migrate the `useEnhancedColorsSetting` to the new `ColorScheme` setting.
+                TryToMigrate();
+
+                QueueColorSchemeUpdate();
+            }
+        }
+
+        public void SubjectBuffersDisconnected(IWpfTextView textView, ConnectionReason reason, Collection<ITextBuffer> subjectBuffers)
+        {
```

Comment:
could add explicit comment that this is implementing the interface and nothing to do